<a href="https://colab.research.google.com/github/Blevins05/data-ai-summer2026/blob/main/analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Analisis de un dataset de partidos de LaLiga 2019-2025

Repasando análisis de datos básico con pandas, y luego visualización con matplotlib y seaborn. La base del análisis de datos.

Mediante `read_csv()` podemos cargar un CSV en pandas.

También podemos trabajar con `
.xlxx` o `.tsv`

In [1]:
import pandas as pd
data = pd.read_csv("matches_laliga.csv")

Mediante el método `head()`, muestro las primeras 5 instancias del dataset de partidos.

Existe igualmente `tail()`, que muestra las 5 últimas.

In [2]:
data.head()

,Unnamed: 0,date,time,comp,round,day,venue,result,gf,ga,...,match report,notes,sh,sot,dist,fk,pk,pkatt,season,team
0,0,2025-08-16,19:30,La Liga,Matchweek 1,Sat,Away,W,3.0,0.0,...,Match Report,NaN,24.0,8.0,18.9,1.0,0,0,2025,Barcelona
1,1,2025-08-23,21:30,La Liga,Matchweek 2,Sat,Away,W,3.0,2.0,...,Match Report,NaN,26.0,10.0,17.0,1.0,0,0,2025,Barcelona
2,2,2025-08-31,21:30,La Liga,Matchweek 3,Sun,Away,D,1.0,1.0,...,Match Report,NaN,11.0,2.0,20.3,0.0,1,1,2025,Barcelona
3,3,2025-09-14,21:00,La Liga,Matchweek 4,Sun,Home,W,6.0,0.0,...,Match Report,NaN,24.0,10.0,18.4,0.0,0,0,2025,Barcelona
4,5,2025-09-21,21:00,La Liga,Matchweek 5,Sun,Home,W,3.0,0.0,...,Match Report,NaN,16.0,7.0,18.2,2.0,0,0,2025,Barcelona


In [3]:
data.tail()

,Unnamed: 0,date,time,comp,round,day,venue,result,gf,ga,...,match report,notes,sh,sot,dist,fk,pk,pkatt,season,team
4695,48,2020-07-05,17:00,La Liga,Matchweek 34,Sun,Home,L,0.0,1.0,...,Match Report,NaN,7.0,2.0,18.9,1.0,0,0,2019,Espanyol
4696,49,2020-07-08,22:00,La Liga,Matchweek 35,Wed,Away,L,0.0,1.0,...,Match Report,NaN,11.0,2.0,18.1,1.0,0,0,2019,Espanyol
4697,50,2020-07-12,14:00,La Liga,Matchweek 36,Sun,Home,L,0.0,2.0,...,Match Report,NaN,14.0,3.0,20.8,3.0,0,0,2019,Espanyol
4698,51,2020-07-16,21:00,La Liga,Matchweek 37,Thu,Away,L,0.0,1.0,...,Match Report,NaN,18.0,6.0,21.3,3.0,0,0,2019,Espanyol
4699,52,2020-07-19,21:00,La Liga,Matchweek 38,Sun,Home,D,0.0,0.0,...,Match Report,NaN,8.0,1.0,17.2,0.0,0,0,2019,Espanyol


En los notebooks de Jupyter, los DataFrames de Pandas se muestran usando las tablas vistas en la celda anterior.

En este caso cada fila corresponde con un cliente, una **instancia**, y las columnas son los **descriptores** de dicha instancia.

Vamos ahora a ver las dimensiones de nuestros datos, los nombres de los descriptores, y los tipos de los descriptores.

La siguiente función nos muestra la dimensión del dataset.

In [4]:
print(data.shape)

(4700, 29)


El dataset de fútbol tiene 4700 instancias, es decir se han jugado 4700 partidos de Primera División Española desde 2019. Para cada partido tenemos 29 descriptores.

El atributo `columns` sirve para mostrar los nombres de las columnas o descriptores

In [5]:
print(data.columns)

Index(['Unnamed: 0', 'date', 'time', 'comp', 'round', 'day', 'venue', 'result',
       'gf', 'ga', 'opponent', 'xg', 'xga', 'poss', 'attendance', 'captain',
       'formation', 'opp formation', 'referee', 'match report', 'notes', 'sh',
       'sot', 'dist', 'fk', 'pk', 'pkatt', 'season', 'team'],
      dtype='object')


Por último, el método `info()` ayuda a ver más información general sobre el DataFrame

In [6]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4700 entries, 0 to 4699
Data columns (total 29 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     4700 non-null   int64  
 1   date           4700 non-null   object 
 2   time           4700 non-null   object 
 3   comp           4700 non-null   object 
 4   round          4700 non-null   object 
 5   day            4700 non-null   object 
 6   venue          4700 non-null   object 
 7   result         4700 non-null   object 
 8   gf             4700 non-null   float64
 9   ga             4700 non-null   float64
 10  opponent       4700 non-null   object 
 11  xg             4700 non-null   float64
 12  xga            4700 non-null   float64
 13  poss           4700 non-null   float64
 14  attendance     3724 non-null   float64
 15  captain        4700 non-null   object 
 16  formation      4700 non-null   object 
 17  opp formation  4700 non-null   object 
 18  referee 

Se puede ver cada uno de los descriptores y el tipo de dato que son.

Es posible transformar el tipo de dato de un descriptor mediante el método `astype()`

Por ejemplo: (aunque no hace falta)

In [7]:
data['gf'] = data['gf'].astype('float')

In [8]:
print(data['gf'])

0       3.0
1       3.0
2       1.0
3       6.0
4       3.0
       ... 
4695    0.0
4696    0.0
4697    0.0
4698    0.0
4699    0.0
Name: gf, Length: 4700, dtype: float64


El método `describe()` aporta estadísticas básicas de cada descriptor numérico. En concreto, el número de nulos, la media, la desv. típica, el rango (mediante los valores min, max), la mediana indicada con el cuartil 50 y los cuartiles .25 y .75

In [9]:
data.describe()

,Unnamed: 0,gf,ga,xg,xga,poss,attendance,notes,sh,sot,dist,fk,pk,pkatt,season
count,4700.000000,4700.000000,4700.000000,4700.000000,4700.000000,4700.000000,3724.000000,0.0,4700.000000,4700.000000,4697.000000,4700.000000,4700.000000,4700.000000,4700.000000
mean,22.061702,1.273617,1.273617,1.264532,1.264532,50.001064,28097.842105,NaN,11.550213,3.804468,18.154652,0.443191,0.130426,0.168936,2021.604255
std,14.395784,1.165048,1.165048,0.776251,0.776251,11.509529,18104.760706,NaN,4.906037,2.296370,3.386085,0.688555,0.362374,0.415661,1.784514
min,0.000000,0.000000,0.000000,0.000000,0.000000,18.000000,13.000000,NaN,0.000000,0.000000,4.800000,0.000000,0.000000,0.000000,2019.000000
25%,9.000000,0.000000,0.000000,0.700000,0.700000,42.000000,14266.000000,NaN,8.000000,2.000000,15.900000,0.000000,0.000000,0.000000,2020.000000
50%,22.000000,1.000000,1.000000,1.100000,1.100000,50.000000,20298.500000,NaN,11.000000,3.000000,18.000000,0.000000,0.000000,0.000000,2022.000000
75%,34.000000,2.000000,2.000000,1.700000,1.700000,58.000000,39957.000000,NaN,14.000000,5.000000,20.000000,1.000000,0.000000,0.000000,2023.000000
max,60.000000,7.000000,7.000000,5.900000,5.900000,82.000000,95745.000000,NaN,40.000000,17.000000,46.200000,5.000000,3.000000,3.000000,2025.000000


Para visualizar stats de descriptores no numéricos, hay que indicar los tipos de datos que interesa mirar en el parámetro ´include´

In [10]:
data.describe(include=['object'])

,date,time,comp,round,day,venue,result,opponent,captain,formation,opp formation,referee,match report,team
count,4700,4700,4700,4700,4700,4700,4700,4700,4700,4700,4700,4700,4700,4700
unique,839,21,1,38,7,2,3,28,245,22,21,36,1,28
top,2025-05-18,21:00,La Liga,Matchweek 1,Sun,Away,W,Valencia,Djené,4-4-2,4-4-2,César Soto,Match Report,Barcelona
freq,20,1328,4700,140,1770,2350,1712,235,183,1298,1298,252,4700,235


Para descriptores categóricos (de tipo `object`) y booleanos (tipo `bool`), podemos usar el método `value_counts`.

In [11]:
data['formation'].value_counts()

,count
formation,
4-4-2,1298
4-2-3-1,1194
4-3-3,750
4-1-4-1,321
5-3-2,277
3-4-3,219
5-4-1,127
3-5-2,98
4-1-3-2,98


La formación más usada durante estos años ha sido una 4-4-2, seguida de cerca por el 4-2-3-1

Si quisieramos visualizarlo en porcentajes, utilizariamos el parámetro `normalize=True`

In [12]:
data['formation'].value_counts(normalize=True)

,proportion
formation,
4-4-2,0.276170
4-2-3-1,0.254043
4-3-3,0.159574
4-1-4-1,0.068298
5-3-2,0.058936
3-4-3,0.046596
5-4-1,0.027021
3-5-2,0.020851
4-1-3-2,0.020851


Casi un 28% de los partidos emplea una formación 4-4-2

Los dataframes los podemos ordenar también mediante 1 de sus descriptores o mediante varios. Esto a su vez se puede hacer de manera ascendiente o descendiente.

Empleamos 2 cosas: el método `sort_values`, y el parámetro `ascending`

Ordeno por número de goles

In [13]:
data.sort_values(by='poss', ascending=False)

,Unnamed: 0,date,time,comp,round,day,venue,result,gf,ga,...,match report,notes,sh,sot,dist,fk,pk,pkatt,season,team
2437,23,2021-12-19,21:00,La Liga,Matchweek 18,Sun,Home,D,0.0,0.0,...,Match Report,NaN,36.0,9.0,19.0,2.0,0,0,2021,Real Madrid
3265,14,2020-12-05,21:00,La Liga,Matchweek 12,Sat,Away,L,1.0,2.0,...,Match Report,NaN,21.0,8.0,20.6,4.0,0,0,2020,Barcelona
3278,35,2021-02-21,14:00,La Liga,Matchweek 24,Sun,Home,D,1.0,1.0,...,Match Report,NaN,19.0,3.0,19.4,0.0,1,1,2020,Barcelona
3288,48,2021-04-29,19:00,La Liga,Matchweek 33,Thu,Home,L,1.0,2.0,...,Match Report,NaN,16.0,4.0,17.1,1.0,0,0,2020,Barcelona
3997,26,2020-01-19,21:00,La Liga,Matchweek 20,Sun,Home,W,1.0,0.0,...,Match Report,NaN,18.0,6.0,21.2,3.0,0,0,2019,Barcelona
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
895,36,2025-05-03,21:00,La Liga,Matchweek 34,Sat,Home,L,1.0,2.0,...,Match Report,NaN,5.0,4.0,11.6,0.0,0,0,2024,Valladolid
99,1,2025-08-23,21:30,La Liga,Matchweek 2,Sat,Home,L,2.0,3.0,...,Match Report,NaN,7.0,4.0,13.5,0.0,1,1,2025,Levante
3621,26,2021-02-21,14:00,La Liga,Matchweek 24,Sun,Away,D,1.0,1.0,...,Match Report,NaN,2.0,0.0,19.4,0.0,1,1,2020,Cadiz
3609,11,2020-12-05,21:00,La Liga,Matchweek 12,Sat,Home,W,2.0,1.0,...,Match Report,NaN,7.0,3.0,15.8,1.0,0,0,2020,Cadiz


El equipo con más posesión en un partido, es el Real Madrid con un 82% de posesión, frente al Cádiz, el 19 de diciembre de la temporada 21/22 que acabo 0-0.

También se puede ordenar por múltiples descriptores.

In [14]:
data.sort_values(by=["poss", 'gf'], ascending=[False, False])

,Unnamed: 0,date,time,comp,round,day,venue,result,gf,ga,...,match report,notes,sh,sot,dist,fk,pk,pkatt,season,team
3265,14,2020-12-05,21:00,La Liga,Matchweek 12,Sat,Away,L,1.0,2.0,...,Match Report,NaN,21.0,8.0,20.6,4.0,0,0,2020,Barcelona
2437,23,2021-12-19,21:00,La Liga,Matchweek 18,Sun,Home,D,0.0,0.0,...,Match Report,NaN,36.0,9.0,19.0,2.0,0,0,2021,Real Madrid
1,1,2025-08-23,21:30,La Liga,Matchweek 2,Sat,Away,W,3.0,2.0,...,Match Report,NaN,26.0,10.0,17.0,1.0,0,0,2025,Barcelona
173,54,2025-05-03,21:00,La Liga,Matchweek 34,Sat,Away,W,2.0,1.0,...,Match Report,NaN,24.0,6.0,17.4,3.0,0,0,2024,Barcelona
3278,35,2021-02-21,14:00,La Liga,Matchweek 24,Sun,Home,D,1.0,1.0,...,Match Report,NaN,19.0,3.0,19.4,0.0,1,1,2020,Barcelona
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
895,36,2025-05-03,21:00,La Liga,Matchweek 34,Sat,Home,L,1.0,2.0,...,Match Report,NaN,5.0,4.0,11.6,0.0,0,0,2024,Valladolid
3621,26,2021-02-21,14:00,La Liga,Matchweek 24,Sun,Away,D,1.0,1.0,...,Match Report,NaN,2.0,0.0,19.4,0.0,1,1,2020,Cadiz
4187,19,2020-01-19,21:00,La Liga,Matchweek 20,Sun,Away,L,0.0,1.0,...,Match Report,NaN,5.0,1.0,26.1,1.0,0,0,2019,Granada
3609,11,2020-12-05,21:00,La Liga,Matchweek 12,Sat,Home,W,2.0,1.0,...,Match Report,NaN,7.0,3.0,15.8,1.0,0,0,2020,Cadiz


Así a primera vista, parece que el Cádiz es el equipo con menor posesión de la competición cuando juega en la 1era división española de fútbol.

Los dataframes se pueden indexar de distinta manera. Con esto, podemos obtener filas únicas, es como hacer queries.

Por ejemplo, la media de goles a favor por partido:

In [15]:
data['gf'].mean()

np.float64(1.2736170212765958)

En 5 años, el promedio de goles por partido es de 1.27 goles por partido

In [16]:
data['referee'].value_counts()

,count
referee,
César Soto,252
Ricardo de Burgos,240
Jesús Gil,240
Pablo González,238
José Luis Munuera,236
Guillermo Cuadra,230
Alejandro Hernández,228
Juan Martínez,218
Mario Melero,216


El colegiado que más partidos ha arbitrado en Primera División durante 5 temporadas (2019-2025), ha sido Cesar Soto, con 252 encuentros disputados. Seguido por De Burgos y Jesus Gil Manzano, con 240.

También esta el indexado condicional, que es lo más parecido a las queries de SQL.

El **indexado condicional** de una columna también es algo muy útil. La sintaxix es `df[P(df['Name'])]`, donde  `P` es alguna condición lógica que es comprobada para cada elemento de la columna `Name`. El resultado de dicho indexado es el Dataframe que consta solo de las filas que satisfacen la condición `P` en la columna `Name`.


Por ejemplo, todos los partidos en los que se han anotado +3'5 goles por parte de algún equipo

In [17]:
data[data['gf'] >= 4].shape


(233, 29)

Otro ejemplo, cual es la posesión promedio de los partidos que terminan 0 a 0, para equipos que juegan en casa

Hay 233 partidos en los que ha habido más de 3 goles por parte de 1 de los 2 equipos

Apunte: En pandas no se puede usar 'and' en indexado condicional, se usa & y cada condición se envuelve en ().


In [18]:
data[(data['gf'] == 0) & (data['ga'] == 0) & (data['venue'] == 'Home')]['poss'].mean()

np.float64(50.83783783783784)

La posesión seria de casi el 51% cuando se juega en casa.

Los dataframes también los podemos indexar por el nombre de la columna (etiqueta) por el nombre de la fila (índice) o por el número de serie de una fila. El método `loc` se usa para indexar por nombre, mientras que el método `iloc()` se utiliza para indexar por número.

Por ejemplo: Los 3 primeros partidos del FC Barcelona

In [19]:
data.loc[data['team'] == 'Barcelona'].iloc[0:3]

,Unnamed: 0,date,time,comp,round,day,venue,result,gf,ga,...,match report,notes,sh,sot,dist,fk,pk,pkatt,season,team
0,0,2025-08-16,19:30,La Liga,Matchweek 1,Sat,Away,W,3.0,0.0,...,Match Report,NaN,24.0,8.0,18.9,1.0,0,0,2025,Barcelona
1,1,2025-08-23,21:30,La Liga,Matchweek 2,Sat,Away,W,3.0,2.0,...,Match Report,NaN,26.0,10.0,17.0,1.0,0,0,2025,Barcelona
2,2,2025-08-31,21:30,La Liga,Matchweek 3,Sun,Away,D,1.0,1.0,...,Match Report,NaN,11.0,2.0,20.3,0.0,1,1,2025,Barcelona


In [20]:
data.loc[0, 'team']          # fila 0, columna 'team'

'Barcelona'

In [21]:
data.loc[0:5]

,Unnamed: 0,date,time,comp,round,day,venue,result,gf,ga,...,match report,notes,sh,sot,dist,fk,pk,pkatt,season,team
0,0,2025-08-16,19:30,La Liga,Matchweek 1,Sat,Away,W,3.0,0.0,...,Match Report,NaN,24.0,8.0,18.9,1.0,0,0,2025,Barcelona
1,1,2025-08-23,21:30,La Liga,Matchweek 2,Sat,Away,W,3.0,2.0,...,Match Report,NaN,26.0,10.0,17.0,1.0,0,0,2025,Barcelona
2,2,2025-08-31,21:30,La Liga,Matchweek 3,Sun,Away,D,1.0,1.0,...,Match Report,NaN,11.0,2.0,20.3,0.0,1,1,2025,Barcelona
3,3,2025-09-14,21:00,La Liga,Matchweek 4,Sun,Home,W,6.0,0.0,...,Match Report,NaN,24.0,10.0,18.4,0.0,0,0,2025,Barcelona
4,5,2025-09-21,21:00,La Liga,Matchweek 5,Sun,Home,W,3.0,0.0,...,Match Report,NaN,16.0,7.0,18.2,2.0,0,0,2025,Barcelona
5,6,2025-09-25,21:30,La Liga,Matchweek 6,Thu,Away,W,3.0,1.0,...,Match Report,NaN,22.0,10.0,16.5,1.0,0,0,2025,Barcelona


Por ejemplo, aqui obtenemos los valores de los primeros 5 partidos en las 8 primeras filas.

In [22]:
data.iloc[0:5, 0:8]

,Unnamed: 0,date,time,comp,round,day,venue,result
0,0,2025-08-16,19:30,La Liga,Matchweek 1,Sat,Away,W
1,1,2025-08-23,21:30,La Liga,Matchweek 2,Sat,Away,W
2,2,2025-08-31,21:30,La Liga,Matchweek 3,Sun,Away,D
3,3,2025-09-14,21:00,La Liga,Matchweek 4,Sun,Home,W
4,5,2025-09-21,21:00,La Liga,Matchweek 5,Sun,Home,W


Para obtener la primera o última instancia de un dataframe, podemos usar respectivamente `df[:1]` y `df[-1:]`.

También se puede aplicar funciones a las celdas, columnas y filas de un dataframe.

Para aplicar funciones a una columna se usa el método `apply()`. Por ejemplo, a continuación mostramos cómo obtener el valor máximo de los distintos descriptores del dataset.

Primero, empezamos importando `numpy` que es una librería de análisis numérico.

Básicamente es una librería para trabajar con arrays multidimensionales de forma muy eficiente. Pandas está construido encima de NumPy.

In [23]:
import numpy as np

Por ejemplo, obtenemos el valor máximo para cada una de las columnas de nuestro dataset

In [24]:
data.apply(np.max)

,0
Unnamed: 0,60
date,2025-09-30
time,22:15
comp,La Liga
round,Matchweek 9
day,Wed
venue,Home
result,W
gf,7.0
ga,7.0


También se puede aplicar a filas, solo hay que utilizar `axis=1`.
Las funciones lambda suelen ser muy útiles en estos casos.

Por ejemplo, partidos donde el resultado empieza por W que son Victorias (no es muy útil pero como ejemplo sirve, ya que con hacer data[data['result'] == 'W'] lo tienes

In [25]:
data[data['result'].apply(lambda x: x.startswith('W'))]

,Unnamed: 0,date,time,comp,round,day,venue,result,gf,ga,...,match report,notes,sh,sot,dist,fk,pk,pkatt,season,team
0,0,2025-08-16,19:30,La Liga,Matchweek 1,Sat,Away,W,3.0,0.0,...,Match Report,NaN,24.0,8.0,18.9,1.0,0,0,2025,Barcelona
1,1,2025-08-23,21:30,La Liga,Matchweek 2,Sat,Away,W,3.0,2.0,...,Match Report,NaN,26.0,10.0,17.0,1.0,0,0,2025,Barcelona
3,3,2025-09-14,21:00,La Liga,Matchweek 4,Sun,Home,W,6.0,0.0,...,Match Report,NaN,24.0,10.0,18.4,0.0,0,0,2025,Barcelona
4,5,2025-09-21,21:00,La Liga,Matchweek 5,Sun,Home,W,3.0,0.0,...,Match Report,NaN,16.0,7.0,18.2,2.0,0,0,2025,Barcelona
5,6,2025-09-25,21:30,La Liga,Matchweek 6,Thu,Away,W,3.0,1.0,...,Match Report,NaN,22.0,10.0,16.5,1.0,0,0,2025,Barcelona
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4665,9,2019-09-15,12:00,La Liga,Matchweek 4,Sun,Away,W,2.0,1.0,...,Match Report,NaN,7.0,2.0,16.8,0.0,0,0,2019,Espanyol
4671,18,2019-10-27,16:00,La Liga,Matchweek 10,Sun,Away,W,1.0,0.0,...,Match Report,NaN,14.0,3.0,17.6,1.0,0,0,2019,Espanyol
4681,31,2020-01-19,16:00,La Liga,Matchweek 20,Sun,Away,W,2.0,1.0,...,Match Report,NaN,16.0,5.0,16.9,0.0,0,0,2019,Espanyol
4684,35,2020-02-09,12:00,La Liga,Matchweek 23,Sun,Home,W,1.0,0.0,...,Match Report,NaN,12.0,7.0,16.8,0.0,0,0,2019,Espanyol


Otro ejemplo:
 Mediante una lambda, y aplicando a filas, podemos clasificar los resultados en puntos:

- W: 3 pts

- D: 1 pts

- L: 0 pts


In [26]:
data['points'] = data['result'].apply(lambda x: 3 if x == 'W' else 1 if x == 'D' else 0)

Además, en este ejemplo hemos creado una nueva entrada en nuestro dataframe, añadiendo el descriptor points o puntos como nueva columna

In [27]:
data.shape

(4700, 30)

Ahora tendríamos 30 descriptores

In [28]:
data['points'].describe()

,points
count,4700.000000
mean,1.364255
std,1.299598
min,0.000000
25%,0.000000
50%,1.000000
75%,3.000000
max,3.000000


El método `map` se puede utilizar para reemplazar valores en una columna pasándole un diccionario de la forma `{old_value: new_value}` como argumento:

Convierto de W/D/L a Victoria/Empate/Derrota

In [29]:
data['result'] = data['result'].map({'W': 'Victoria', 'D': 'Empate', 'L' : 'Derrota'})

In [30]:
data['result'].value_counts()

,count
result,
Victoria,1712
Derrota,1712
Empate,1276


Agrupaciones. Para construir grupos de datos en Pandas, utilizamos generalmente una construcción como la siguiente.

```python
df.groupby(by=grouping_columns)[columns_to_show].function()
```

1. Primero, el método `groupby` divide `grouping_columns` por sus valores, que se convierten en un nuevo índice en el dataframe resultante.
2. Seguidamente, las columnas de interés se seleccionan (`columns_to_show`). Si no se incluye `columns_to_show` se muestran todas las clausulas que no hayan sido agrupadas.
3. Finalmente, una o varias funciones se aplican para obtener los grupos por las columnas seleccionadas.


Por ejemplo, las siguientes agrupaciones son bastante útiles

Media de goles por equipo:

In [31]:
data.groupby('team')['gf'].mean()

,gf
team,
Alaves,0.918782
Almeria,1.210526
Athletic Club,1.272340
Atletico Madrid,1.723404
Barcelona,2.174468
Cadiz,0.835526
Celta Vigo,1.229787
Eibar,0.894737
Elche,0.942149


El Barcelona es el equipo que más goles ha promediado en estas 5 últimas temporadas, con 2.17 goles por partido. Seguido por el Real Madrid, con 2.01

El Real Oviedo es el equipo con menos goles promediado por encuentro que ha pasado por LaLiga en estos últimos 5 años.

Victorias, derrotas y empates por equipo

In [32]:
data.groupby(['team', 'result']).size()

team        result  
Alaves      Derrota      95
            Empate       51
            Victoria     51
Almeria     Derrota      42
            Empate       20
                       ... 
Valladolid  Empate       42
            Victoria     29
Villarreal  Derrota      69
            Empate       59
            Victoria    107
Length: 83, dtype: int64

Puntos totales de cada equipo por temporada.

In [33]:
data.groupby(['team', 'season'])['points'].sum()

team        season
Alaves      2019      39
            2020      38
            2021      31
            2023      46
            2024      42
                      ..
Villarreal  2021      59
            2022      64
            2023      53
            2024      70
            2025      16
Name: points, Length: 140, dtype: int64

Si quisieramos mostrar este último ejemplo completo, podemos utilizar `reset_index` que convierte el índice jerárquico en columnas normales y muestra todo como un DataFrame limpio.

In [34]:
data.groupby(['team', 'season'])['points'].sum().reset_index()

,team,season,points
0,Alaves,2019,39
1,Alaves,2020,38
2,Alaves,2021,31
3,Alaves,2023,46
4,Alaves,2024,42
...,...,...,...
135,Villarreal,2021,59
136,Villarreal,2022,64
137,Villarreal,2023,53
138,Villarreal,2024,70


Estadísticas de posesión y xG por resultado

In [35]:
data.groupby('result')[['poss', 'xg', 'sh']].mean()

,poss,xg,sh
result,,,
Derrota,48.732477,0.956659,10.663551
Empate,50.001567,1.125862,11.328370
Victoria,51.269276,1.675759,12.602220


Se puede hacer algo parecido, pero en este caso pasando una lista de funciones a `agg()`:

In [36]:
data.groupby('team')['gf'].agg(['mean', 'sum', 'max', 'min'])

,mean,sum,max,min
team,,,,
Alaves,0.918782,181.0,4.0,0.0
Almeria,1.210526,92.0,6.0,0.0
Athletic Club,1.272340,299.0,7.0,0.0
Atletico Madrid,1.723404,405.0,7.0,0.0
Barcelona,2.174468,511.0,7.0,0.0
Cadiz,0.835526,127.0,3.0,0.0
Celta Vigo,1.229787,289.0,6.0,0.0
Eibar,0.894737,68.0,3.0,0.0
Elche,0.942149,114.0,4.0,0.0


Las tablas resumen en un dataframe nos sirven para ver como las muestras del mismo se distribuyen en el contexto de 2 variables.

Para esto se utiliza `crosstab()` junto a las 2 columnas que se quiere estudiar

Por ejemplo, cuantas veces cada equipo gana/empata/pierde

In [37]:

pd.crosstab(data['team'], data['result'])


result,Derrota,Empate,Victoria
team,,,
Alaves,95,51,51
Almeria,42,20,14
Athletic Club,70,72,93
Atletico Madrid,41,57,137
Barcelona,37,40,158
Cadiz,64,53,35
Celta Vigo,95,70,70
Eibar,38,21,17
Elche,59,35,27


Algunas preguntas sobre el dataset: Para comprenderlo mejor y poder ver más fácilmente la relación entre sus descriptores


¿Cuál es el equipo con más puntos totales en toda la historia del dataset?

In [85]:
data.groupby('team')['points'].sum().sort_values(ascending=False)

,points
team,
Real Madrid,532
Barcelona,514
Atletico Madrid,468
Villarreal,380
Real Sociedad,362
Sevilla,358
Real Betis,356
Athletic Club,351
Osasuna,300


¿Cuál es la media de goles marcados jugando en casa vs fuera?

In [51]:
data[data['venue']=='Home']['gf'].mean()

np.float64(1.4387234042553192)

In [65]:
data[data['venue']=='Away']['gf'].mean()

np.float64(1.1085106382978724)

Se marcan más goles cuando se juega en casa de promedio, que cuando se juega fuera de casa.

 ¿Qué formación tiene mayor porcentaje de victorias?

In [57]:
data['formation'].value_counts(normalize=True) * 100

,proportion
formation,
4-4-2,27.617021
4-2-3-1,25.404255
4-3-3,15.957447
4-1-4-1,6.829787
5-3-2,5.893617
3-4-3,4.659574
5-4-1,2.702128
3-5-2,2.085106
4-1-3-2,2.085106


In [86]:
data.groupby('formation')['result'].apply(lambda x: (x=='Victoria').mean() * 100)

,result
formation,
3-1-4-2,38.983051
3-2-4-1,100.000000
3-3-3-1,0.000000
3-4-1-2,26.923077
3-4-3,42.009132
3-4-3◆,50.000000
3-5-1-1,16.666667
3-5-2,34.693878
4-1-2-1-2◆,58.823529


¿Cuál es el árbitro con mayor media de goles por partido en los partidos que ha dirigido?

In [61]:
data['referee'].value_counts()

,count
referee,
César Soto,252
Ricardo de Burgos,240
Jesús Gil,240
Pablo González,238
José Luis Munuera,236
Guillermo Cuadra,230
Alejandro Hernández,228
Juan Martínez,218
Mario Melero,216


In [84]:
data.groupby('referee')['gf'].mean().sort_values(ascending=False)

,gf
referee,
Javier Alberola,1.475000
Guillermo Cuadra,1.421739
Ricardo de Burgos,1.408333
Juan Pulido,1.390909
Miguel Ángel Ortiz Arias,1.371795
Valentín Pizarro,1.371429
Alejandro Hernández,1.364035
José Luis Munuera,1.360169
Alejandro Muñíz,1.358974


¿Hay alguna correlación entre la posesión y los puntos obtenidos?

In [63]:
pd.crosstab(data['poss'], data['points'])

points,0,1,3
poss,,,
18.0,0,1,1
19.0,3,1,1
20.0,3,1,0
21.0,1,3,2
22.0,2,1,1
...,...,...,...
78.0,1,1,2
79.0,2,3,1
80.0,0,1,3
